enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+Otherwise the replacement is identical to the RollingUpdate strategy.
+
+#### Protecting etcd quorum during update operations
+
+With the introduction of the [etcd protection enhancement](https://github.com/openshift/enhancements/pull/943), the
+ControlPlaneMachineSet does not need to observe anything etcd related during scaling operations.
+In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
+from being removed until they are no longer needed.
+When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
+Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
+new instances.
+Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
+removed by the Machine controller in the normal way.
+
+#### Removing/disabling the ControlPlaneMachineSet
+
+As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
+`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
+the Control Plane Machines in place.
+
+Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
+instances. This will prevent the garbage collector from removing the Control Plane Machines when the
+`ControlPlaneMachineSet` is deleted.
+
+If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.
+
+#### Installing a ControlPlaneMachineSet within an existing cluster
+
+When adding a ControlPlaneMachineSet to a existing cluster, the end user will need to define the
+`ControlPlaneMachineSet` resource by copying the existing Control Plane Machine ProviderSpecs.
+Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
+field within the ControlPlaneMachineSet spec.
+
+To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook
+that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.
+We will also validate that the failure domains align with those within the cluster already, this will prevent users
+from accidentally migrating their Control Plane Machines from multiple availability zones to a single zone.
+
+If no Control Plane Machines exist, or they are in a non-Running state, the operator will report degraded until this
+issue is resolved. This creates a dependency for the `ControlPlaneMachineSet` operator on the Machine API. It will be
+required to run at a higher run-level than Machine API.
+
+In UPI or misconfigured clusters, a user adding a `ControlPlaneMachineSet` will result in a degraded cluster.
```

Comment:
How is the determination going to be made that the cluster is UPI? UPI is an installation style but results in the same cluster as IPI.

Reply:
I'm using UPI as shorthand here so will update to spell it out, What I actually want to say is in a cluster where it does not have valid Control Plane Machines and Nodes, the cluster will degrade.

If a user is willing to create Control Plane Machines in their UPI cluster and can get them to a Running phase, they can then add a CPMS to take over management, but we don't intend to do that initial bootstrap for them